src/Microsoft.Kusto.ServiceLayer/DataSource/KustoDataSource.cs

```diff
@@ -626,93 +612,311 @@ internal IEnumerable<ColumnInfo> GetColumnMetadata(string databaseName)
+                AddToFolderMetadata(tableFolderKey.ToString(), new List<FolderMetadata> {tableFolder});
+                tableFolderKey.Append($".{tableFolder.Name}");
+                
+                SetFolderMetadataForTables(objectMetadata, columnInfos, tableFolderKey.ToString());
+            }
+            
+            var columnsGroupByTable = columnInfos
+                .Where(row =>
+                    !string.IsNullOrWhiteSpace(row.Table)
+                    && !string.IsNullOrWhiteSpace(row.Name)
+                    && !string.IsNullOrWhiteSpace(row.DataType))
+                .GroupBy(row => row.Table, StringComparer.OrdinalIgnoreCase);
+            
+            SetTableMetadata(objectMetadata.Name, columnInfos, tableFolderKey.ToString());
+            SetColumnMetadata(objectMetadata.Name, columnsGroupByTable);
+        }
+
+        private void SetFolderMetadataForTables(DataSourceObjectMetadata objectMetadata, IEnumerable<ColumnInfo> columnInfos, string tableFolderKey)
+        {
+            var columnsGroupByFolder = columnInfos
+                .GroupBy(x => x.Folder, StringComparer.OrdinalIgnoreCase)
+                .ToList();
+            
+            var tableFolders = new List<FolderMetadata>();
 
-            if(!tableMetadata.ContainsKey(databaseName))
+            foreach (var columnGroup in columnsGroupByFolder)
             {
-                var columns = GetColumnMetadata(databaseName).Materialize();
+                // skip tables without folders
+                if (string.IsNullOrWhiteSpace(columnGroup.Key))
+                {
+                    continue;
+                }
+
+                var folder = DataSourceFactory.CreateFolderMetadata(objectMetadata, tableFolderKey.ToString(), columnGroup.Key);
+                tableFolders.Add(folder);
+            }
+
+            AddToFolderMetadata(tableFolderKey.ToString(), tableFolders);
+        }
+
+        private void AddToFolderMetadata(string folderKey, List<FolderMetadata> folders)
+        {
+            if (_folderMetadata.ContainsKey(folderKey))
+            {
+                folders.AddRange(_folderMetadata[folderKey]);
+            }
+            
+            _folderMetadata[folderKey] = folders.OrderBy(x => x.PrettyName, StringComparer.OrdinalIgnoreCase);
+        }
+
+        private void AddToFunctionMetadata(string functionKey, List<FunctionMetadata> functions)
+        {
+            if (_functionMetadata.ContainsKey(functionKey))
+            {
+                functions.AddRange(_functionMetadata[functionKey]);
+            }
+            
+            _functionMetadata[functionKey] = functions.OrderBy(x => x.PrettyName, StringComparer.OrdinalIgnoreCase);
+        }
+
+        private void LoadFunctionSchema(DataSourceObjectMetadata objectMetadata)
+        {
+            IEnumerable<FunctionInfo> functionInfos = GetFunctionInfos(objectMetadata.Name);
```

Comment:
Shouldn't we first check that the function metadata has already been populated for this object/key. Otherwise we'll keep querying every time we expand/collapse this folder.

Reply:
I added this logic 1 level up where this function gets called.

`/// <inheritdoc/>
        private IEnumerable<DataSourceObjectMetadata> GetDatabaseSchema(DataSourceObjectMetadata objectMetadata, bool includeSizeDetails)
        {
            // Check if the database exists
            ValidationUtils.IsTrue<ArgumentException>(DatabaseExists(objectMetadata.Name).Result, $"Database '{objectMetadata}' does not exist.");

            var allMetadata = GetAllMetadata(objectMetadata.Urn);

            // if the records have already been loaded them return them
            if (allMetadata.Any())
            {
                return allMetadata;
            }

            LoadTableSchema(objectMetadata);
            LoadFunctionSchema(objectMetadata);
}`